### ISSUE TITLE
Consul ESM has dissociative identity disorder?

### ISSUE THREAD

**Issue by ashald** [hashicorp/consul-esm#39]
We runs Consul ESM in 3-node "clusters" in several Consul clusters.

TL;DR this is what we see:

## Sample 1

![image](https://user-images.githubusercontent.com/397987/57542061-8f2fd300-731e-11e9-9751-4ec1ac78dc5d.png)

Here is how the KV looks like in YAML-like representation:
```yaml
KV:
  __esm__:
      leader: "" # part of session '3034f8c2-ddd3-a733-c3a4-a07b173b0ab7' held by host ''***31'
      agents:
        consul-esm:5460c1fa-4368-62c3-311a-0934fa005cdc: {"Nodes":null,"Probes":["esm-test-host-192-10","esm-test-host-192-15","esm-test-host-192-20"]} # entry name corresponds to ESM running on ***33, part of session '3034f8c2-ddd3-a733-c3a4-a07b173b0ab7' owned by host ''***31'
        consul-esm:6b66d99a-f83b-d678-b710-42b5242347af: {"Nodes":null,"Probes":["esm-test-host-192-11","esm-test-host-192-16","esm-test-host-192-21"]} # entry name corresponds to ESM running on ***33, part of session '3034f8c2-ddd3-a733-c3a4-a07b173b0ab7' owned by host ''***31'
        consul-esm:ab78b0e1-5bb7-6114-d758-5a59dc814361: {"Nodes":null,"Probes":["esm-test-host-192-12","esm-test-host-192-17","esm-test-host-192-22"]} # entry name corresponds to ESM running on ***31, part of session '3034f8c2-ddd3-a733-c3a4-a07b173b0ab7' owned by host ''***31'
        consul-esm:ec9c23a5-b092-295b-76d6-dc7f0508da69: {"Nodes":null,"Probes":["esm-test-host-192-13","esm-test-host-192-18","esm-test-host-192-23"]} # entry name corresponds to ESM running on ***32, part of session '3034f8c2-ddd3-a733-c3a4-a07b173b0ab7' owned by host ''***31'
        consul-esm:f2001c4c-07ce-7a91-bcf3-67c168c9236d: {"Nodes":null,"Probes":["esm-test-host-192-14","esm-test-host-192-19"]}                        # entry name corresponds to ESM running on ***32, part of session '3034f8c2-ddd3-a733-c3a4-a07b173b0ab7' owned by host ''***31'
      probes:
        esm-test-host-192-10: AQAAAA7UZ58zIPqZFf//
        esm-test-host-192-12: AQAAAA7UZ580Dz4bBf//
        esm-test-host-192-14: AQAAAA7UZ58tMd74BP//
        esm-test-host-192-15: AQAAAA7UZ58vMuHD9///
        esm-test-host-192-17: AQAAAA7UZ58tJvKprv//
        esm-test-host-192-19: AQAAAA7UZ58yLAwK2f//
        esm-test-host-192-20: AQAAAA7UZ58pCS3VxP//
        esm-test-host-192-22: AQAAAA7UZ58xCQc+a///
        

```
Node status:
<img width="1611" alt="image" src="https://user-images.githubusercontent.com/397987/57543750-3c0c4f00-7323-11e9-8deb-e6d0ca9bffd8.png">

## Sample 2

<img width="1616" alt="image" src="https://user-images.githubusercontent.com/397987/57542007-67d90600-731e-11e9-8838-014911969e0d.png">

Here is how the KV looks like in YAML-like representation:

```yaml
KV:
  __esm__:
      leader: "" # part of session 'edc63fda-a348-f7a2-bdd5-749cf57751f1' held by host ''***11'
      agents:
        consul-esm:556c6406-b7fe-199f-716c-6593e2512d46: {"Nodes":null,"Probes":["esm-test-host-190-10","esm-test-host-190-14","esm-test-host-190-18"]} # entry name corresponds to ESM running on ***12
        consul-esm:a73bd45f-8ddd-26e1-2e0d-9ffa62306be5: {"Nodes":null,"Probes":["esm-test-host-190-11","esm-test-host-190-15","esm-test-host-190-19"]} # entry name corresponds to ESM running on ***21
        consul-esm:bedfcb00-5198-4e4c-b644-9c6df993a222: {"Nodes":null,"Probes":["esm-test-host-190-12","esm-test-host-190-16","esm-test-host-190-20"]} # entry name corresponds to ESM running on ***11
        consul-esm:d1da7756-897e-c35b-7bb1-a9bd8cd34705: {"Nodes":null,"Probes":["esm-test-host-190-13","esm-test-host-190-17","esm-test-host-190-21"]} # entry name corresponds to ESM running on ***11
```

When we just registered external services, a week or few ago, 3 oof them that correspond to:
```yaml
consul-esm:d1da7756-897e-c35b-7bb1-a9bd8cd34705: {"Nodes":null,"Probes":["esm-test-host-190-13","esm-test-host-190-17","esm-test-host-190-21"]} # entry name corresponds to ESM running on ***11
```
we're missing info about the status but after a while became "green":
<img width="1621" alt="image" src="https://user-images.githubusercontent.com/397987/57543550-ae306400-7322-11e9-9406-e7d2424f6a73.png">

# Conclusion
Not really sure what it all means but my gut feeling is that it's not normal. We would be looking forward to getting this addressed and would appeciate an advice on how we can get it to a "normal state". Thanks!

**Comment by ashald** [comment]
The `Sample 1` from above has evolved into following:
<img width="1608" alt="image" src="https://user-images.githubusercontent.com/397987/58899002-e3359980-86c9-11e9-9aa6-03277878e044.png">

Which suggests that the issue is not a one-off configuration error but rather some sort of a bug that manifests more and more over time.

**Comment by ashald** [comment]
After digging throughout Consul's catalog I figured out few things.

1) To my surprise, ESM on its own is not registered as an agent's service but rather as a Catalog service. I might be missing something obvious about the design principles behind the implementation but this was something unexpected for me. This effectively means that there are no checks associated with each of ESM instances and even if ESM node goes down, unless it exits cleanly and deregisters properly (which we may never assume), it's going to be all green in the Catalog.

2) To even bigger surprise, the "zombie" ESMs have their KV dirs in place. Once again, this probably meant to be cleaned up automatically but somehow did not happen. It's hard to make a guess here and we probably need to check the source code to understand the issue better but here are few observations. In `Sample 2` above all "esm instance dirs" are linked to a Consul session which I assumed is how they supposed to be cleaned up - once ESM instance exits its session is invalidated and all linked KV entries are removed. As shown in the original post, something seems to be off here as *all* of the entries are part of 1 session. I either don't understand the design behind this behavior or something is off here. Last but not least, in`Sample 1` above none of "esm instance dirs in KV" belongs to any sessions.

3) It appears that Consul-ESM is "stateless" as in it doesn't persist any data on disk. Neither it accepts "esm node id" as part of its config. It means that each time it restarts it generates a new UUID for its ID and registers with it anew. If previous instance was not deregistered properly - it will still remain in the Catalog as a "zombie" and the new instance will register along.

I'm very light on how exactly ESM works internally but from my perspective following things have to be considered as potential ways to improve things:
1) Persist ESM's node ID and/or accept it via config
2) Consider registering ESM on its own as an agent service rather than a Catalog service [so that it can have its own health-check].

**Comment by ashald** [comment]
Correction to (1) in previous comment - it seems that some of the ESM instances registered as "agent services" but all "zombies" are only present in the "catalog". At this point I'm not really sure what's going on here. 🙃

**Comment by ashald** [comment]
Ok, I was able to reproduce the issue:
```
sudo kill -9 $(pidof consul-esm)
sudo service consul-esm restart
```

This is the "artificial case" but I believe something like that can happen easily, e.g. in case of a host restart.

**Comment by ashald** [comment]
It appears that issues is reproducible with any recent version of Consul:

### 1. Start Consul Server
In a separate terminal:
```bash
$ docker run --rm -p 0.0.0.0:8500:8500 -e CONSUL_LOCAL_CONFIG='{"acl_datacenter": "dc1", "acl_master_token": "secret", "bootstrap_expect": 1, "server": true, "ui": true}' 'consul:latest'
```

### 2. Create ESM Config
```bash
$ cat > esm.json << EOF
{
    "consul_kv_path": "__esm__/",
    "ping_type": "socket",
    "token": "secret"
}
EOF
```

### 3. Start Consul-ESM
```
$ consul-esm -config-file=esm.json
2019/06/07 03:26:30 [INFO] Connecting to Consul on 127.0.0.1:8500...
Consul ESM running!
            Datacenter: (default)
               Service: "consul-esm"
           Service Tag: ""
            Service ID: "consul-esm:6521bccc-a66a-6d45-b085-5f1529620de4"
Node Reconnect Timeout: "72h0m0s"

Log data will now stream in as it occurs:

2019/06/07 03:26:30 [INFO] Trying to obtain leadership...
2019/06/07 03:26:30 [INFO] Obtained leadership
2019/06/07 03:26:31 [INFO] Rebalanced 0 external nodes across 1 ESM instances
```

### 4. Observe single ESM node
<img width="1627" alt="image" src="https://user-images.githubusercontent.com/397987/59079625-9e208b80-88b2-11e9-9f62-956055582d88.png">

<img width="440" alt="image" src="https://user-images.githubusercontent.com/397987/59079646-af699800-88b2-11e9-9d75-ef98af0f29ed.png">


### 5. Kill -9 Consul ESM
```
$ kill -9 $(pidof consul-esm)

### 6. Observe Consul-ESM remain in Consul
<img width="1616" alt="image" src="https://user-images.githubusercontent.com/397987/59079732-125b2f00-88b3-11e9-978b-7d347c9ed1e3.png">

```bash
curl http://localhost:8500/v1/agent/services
{
    "consul-esm:6521bccc-a66a-6d45-b085-5f1529620de4": {
        "ID": "consul-esm:6521bccc-a66a-6d45-b085-5f1529620de4",
        "Service": "consul-esm",
        "Tags": [],
        "Meta": {},
        "Port": 0,
        "Address": "",
        "Weights": {
            "Passing": 1,
            "Warning": 1
        },
        "EnableTagOverride": false
    }
}
```

```bash
$ curl http://localhost:8500/v1/catalog/service/consul-esm
[
    {
        "ID": "cfe2faa2-b320-2f0b-cd3f-3ef4c11e7118",
        "Node": "6724a586479c",
        "Address": "127.0.0.1",
        "Datacenter": "dc1",
        "TaggedAddresses": {
            "lan": "127.0.0.1",
            "wan": "127.0.0.1"
        },
        "NodeMeta": {
            "consul-network-segment": ""
        },
        "ServiceKind": "",
        "ServiceID": "consul-esm:6521bccc-a66a-6d45-b085-5f1529620de4",
        "ServiceName": "consul-esm",
        "ServiceTags": [],
        "ServiceAddress": "",
        "ServiceWeights": {
            "Passing": 1,
            "Warning": 1
        },
        "ServiceMeta": {},
        "ServicePort": 0,
        "ServiceEnableTagOverride": false,
        "ServiceProxyDestination": "",
        "ServiceProxy": {},
        "ServiceConnect": {},
        "CreateIndex": 15,
        "ModifyIndex": 15
    }
]
```
```bash
$ curl "http://localhost:8500/v1/kv/?recurse=true"
[
    {
        "LockIndex": 0,
        "Key": "__esm__/agents/consul-esm:6521bccc-a66a-6d45-b085-5f1529620de4",
        "Flags": 0,
        "Value": "eyJOb2RlcyI6bnVsbCwiUHJvYmVzIjpudWxsfQ==",
        "CreateIndex": 19,
        "ModifyIndex": 19
    },
    {
        "LockIndex": 1,
        "Key": "__esm__/leader",
        "Flags": 3304740253564472344,
        "Value": null,
        "CreateIndex": 18,
        "ModifyIndex": 36
    }
]
```

### 6. Start Consul ESM again
```
$ consul-esm -config-file=esm.json
2019/06/07 03:34:15 [INFO] Connecting to Consul on 127.0.0.1:8500...
Consul ESM running!
            Datacenter: (default)
               Service: "consul-esm"
           Service Tag: ""
            Service ID: "consul-esm:5f86dc26-0817-1c64-dd8a-9bcfb7b56532"
Node Reconnect Timeout: "72h0m0s"

Log data will now stream in as it occurs:

2019/06/07 03:34:15 [INFO] Trying to obtain leadership...
2019/06/07 03:34:15 [INFO] Obtained leadership
2019/06/07 03:34:16 [INFO] Rebalanced 0 external nodes across 1 ESM instances
```

### 7. Observe extra entry in Consul
<img width="1616" alt="image" src="https://user-images.githubusercontent.com/397987/59079863-bcd35200-88b3-11e9-90ed-33e0c5c9cc96.png">

```bash
$ curl http://localhost:8500/v1/agent/services
{
    "consul-esm:5f86dc26-0817-1c64-dd8a-9bcfb7b56532": {
        "ID": "consul-esm:5f86dc26-0817-1c64-dd8a-9bcfb7b56532",
        "Service": "consul-esm",
        "Tags": [],
        "Meta": {},
        "Port": 0,
        "Address": "",
        "Weights": {
            "Passing": 1,
            "Warning": 1
        },
        "EnableTagOverride": false
    },
    "consul-esm:6521bccc-a66a-6d45-b085-5f1529620de4": {
        "ID": "consul-esm:6521bccc-a66a-6d45-b085-5f1529620de4",
        "Service": "consul-esm",
        "Tags": [],
        "Meta": {},
        "Port": 0,
        "Address": "",
        "Weights": {
            "Passing": 1,
            "Warning": 1
        },
        "EnableTagOverride": false
    }
}
```

```bash
$ curl http://localhost:8500/v1/catalog/service/consul-esm
[
    {
        "ID": "cfe2faa2-b320-2f0b-cd3f-3ef4c11e7118",
        "Node": "6724a586479c",
        "Address": "127.0.0.1",
        "Datacenter": "dc1",
        "TaggedAddresses": {
            "lan": "127.0.0.1",
            "wan": "127.0.0.1"
        },
        "NodeMeta": {
            "consul-network-segment": ""
        },
        "ServiceKind": "",
        "ServiceID": "consul-esm:5f86dc26-0817-1c64-dd8a-9bcfb7b56532",
        "ServiceName": "consul-esm",
        "ServiceTags": [],
        "ServiceAddress": "",
        "ServiceWeights": {
            "Passing": 1,
            "Warning": 1
        },
        "ServiceMeta": {},
        "ServicePort": 0,
        "ServiceEnableTagOverride": false,
        "ServiceProxyDestination": "",
        "ServiceProxy": {},
        "ServiceConnect": {},
        "CreateIndex": 51,
        "ModifyIndex": 51
    },
    {
        "ID": "cfe2faa2-b320-2f0b-cd3f-3ef4c11e7118",
        "Node": "6724a586479c",
        "Address": "127.0.0.1",
        "Datacenter": "dc1",
        "TaggedAddresses": {
            "lan": "127.0.0.1",
            "wan": "127.0.0.1"
        },
        "NodeMeta": {
            "consul-network-segment": ""
        },
        "ServiceKind": "",
        "ServiceID": "consul-esm:6521bccc-a66a-6d45-b085-5f1529620de4",
        "ServiceName": "consul-esm",
        "ServiceTags": [],
        "ServiceAddress": "",
        "ServiceWeights": {
            "Passing": 1,
            "Warning": 1
        },
        "ServiceMeta": {},
        "ServicePort": 0,
        "ServiceEnableTagOverride": false,
        "ServiceProxyDestination": "",
        "ServiceProxy": {},
        "ServiceConnect": {},
        "CreateIndex": 15,
        "ModifyIndex": 15
    }
]
```

```bash
$ curl "http://localhost:8500/v1/kv/?recurse=true"
[
    {
        "LockIndex": 0,
        "Key": "__esm__/agents/consul-esm:5f86dc26-0817-1c64-dd8a-9bcfb7b56532",
        "Flags": 0,
        "Value": "eyJOb2RlcyI6bnVsbCwiUHJvYmVzIjpudWxsfQ==",
        "CreateIndex": 55,
        "ModifyIndex": 55
    },
    {
        "LockIndex": 2,
        "Key": "__esm__/leader",
        "Flags": 3304740253564472344,
        "Value": null,
        "Session": "8f5ab945-e17f-6438-0576-596c62771237",
        "CreateIndex": 18,
        "ModifyIndex": 54
    }
]
```

### 8. Repeat
Steps 3-7 can be repeated indefinitely.
In order to observe extra entries in KV one has to start more than 1 ESM in parallel and then `kill -9` one of them and restart.

**Comment by lornasong** [comment]
Hi @ashald! Apologies for the delayed reply. Thanks for your issue and the really clear reproducing steps! To summarize, I believe the issue you are experiencing is that when consul-esm is killed, it is not deregistered and remains a zombie.

Using your steps, I was able to reproduce the same behavior where upon a `kill -9`, the consul-esm is killed but it remains registered with consul with a failing service check.

It looks like this is expected behavior - at the time of writing, when consul-esm registers itself with consul, it registers with `DeregisterCriticalServiceAfter` = 30 minutes. While reproducing your steps, after waiting 30 minutes after `kill -9`, I was able to see that the zombie consul-esm was deregistered in the consul log output.

If you’re interested, here’s the code for the variable that is used to set the deregister time:
```
// deregisterTime is the time the TTL check must be in a failed state for
// the ESM service in consul to be deregistered.
deregisterTime = 30 * time.Minute
```

I’ll add this clarification to the readme for future people who are caught off guard by this as well.

Hope this helps. Thanks!

**Comment by edio** [comment]
@lornasong , coincidentally I started debugging an issue we have with consul esm just today, and I think this issue deserves much more attention and perhaps should be considered critical.

-------------

First of all, let me reiterate, on how _consul-esm_ handles health checks.

As far as I understand, _consul-esm_ does sharding of all health checks between all registered _consul-esm_ instances. If we have 5 instances of _consul-esm_ in the catalog and 20 health checks, every _consul-esm_ isntance will receive ~4 health checks to run. 

If I picked up terminology from the source code correctly, this is called _coordination_.

If we have a dead _consul-esm_ instance in catalog, then health checks assigned to it will not run.

-------------

> It looks like this is expected behavior - at the time of writing, when consul-esm registers itself with consul, it registers with DeregisterCriticalServiceAfter = 30 minutes. While reproducing your steps, after waiting 30 minutes after kill -9

The scenario @ashald  has offered includes restarting of the service after it is killed. This means, that zombie _consul-esm_ health check will start passing again after we restart _consul-esm_ on that node.

Upon restart _consul-esm_, as was pointed out, will generate new ID for itself. After _coordination_ the health checks will be sharded between zombie and real _consul-esm_ instances. The restarted process will only pick health checks that are assigned to its ID however.

Thus we'll end up with bunch of health checks that are never executed.

-------------

This is a different issue, but health checks for _consul-esm_ should be smarter too. Try running mulitple _consul-esm_ instances with 2 different tokens. Then revoke one of the token.
At this point half of the _consul-esm_ isntances will stop working, but health checks will never fail.

**Comment by lornasong** [comment]
Hi @edio, thanks for your comment and all of the information you provided. It looks like you have two areas of concern: (1) dead consul-esm instances will lead to unexecuted health checks (2) revoking tokens on consul-esm instances will lead to unexecuted health checks. Please correct if I misunderstood. I’ll respond in order below.

(1)

> consul-esm does sharding of all health checks between all registered consul-esm instances

To clarify: consul-esm splits all the health checks between all registered _healthy_ consul-esm instances, where ‘healthy’ means the instance is passing its TTL check.

This means:
- if we have healthy consul-esm instances and a dead consul-esm, then the health checks assigned to the dead instance will be reassigned to the healthy ones during the next splitting cycle, approximately after 5s and not after the 30 min deregistration wait
- new health checks will be assigned to healthy consul-esm instances rather than dead instances
- if there are no healthy consul-esm instances, then no health checks will be run. This is unfortunate however not unexpected behavior.

You mentioned:
> in our prod environment we observe healthy "zombie" consul-esm instances. I.e. they take part in coordination and get health checks assigned to them

Would you be able to clarify what you mean by “zombie”? My understanding of “zombie” from ashald's comment was an instance that is dead  — e.g. killed by a `sudo kill -9` — but not yet deregistered. For example, when I test a cluster of healthy and dead (but not yet deregistered) instances, I observe the behavior described in the bullets above. Am I misunderstanding you?

Would you be also mind doing a `curl <insert-base-url>/v1/health/checks/consul-esm` and return the output for your zombie instance? For example, here’s my dead instance:

```
{
	"Node": "agent-one",
	"CheckID": "consul-esm:a66fb414-a865-7937-0227-9c0aabc46521:agent-ttl",
	"Name": "Consul External Service Monitor Alive",
	"Status": "critical",
	"Notes": "This check is periodically updated as long as the agent is alive.",
	"Output": "TTL expired",
	"ServiceID": "consul-esm:a66fb414-a865-7937-0227-9c0aabc46521",
	"ServiceName": "consul-esm",
	"ServiceTags": [],
	"Type": "ttl",
	"Definition": {},
	"CreateIndex": 851,
	"ModifyIndex": 985
}
```

(2)

> Try running mulitple consul-esm instances with 2 different tokens. Then revoke one of the token.
> At this point half of the consul-esm isntances will stop working, but health checks will never fail.

I was able to reproduce your issue. On revoking one consul-esm’s token, I observed that:
- revoked-token consul-esm was still marked healthy/passing in API request to `v1/health/node/:node`
- revoked-token consul-esm’s assigned health-checks were not redistributed to the healthy, un-revoked consul-esm
- 30 minutes after I revoked the token, the revoked-token consul-esm attempted and failed to deregister with a log message similar to that in UPDATE 2

This is happening because the token used to register consul-esm is also the token needed to sync consul-esm. Hence, consul-esm’s health status doesn’t get updated and it is unable to deregister.

This is in an existing issue in consul: https://github.com/hashicorp/consul/issues/4372 - there’s more helpful details there. It looks like the recommended solution described in the issue is to re-register the application with the new token. Given that consul-esm re-registers with a new ID each time, this solution will lead to dead orphaned consul-esm instances which hold onto their assigned external health checks - as you described.

I went ahead and opened a separate issue for this: https://github.com/hashicorp/consul-esm/issues/53. If you see anything is missing or misrepresented, please feel free to comment there. Thanks for raising this issue!

**Comment by edio** [comment]
> Would you be able to clarify what you mean by “zombie”? My understanding of “zombie” from ashald's comment was an instance that is dead — e.g. killed by a sudo kill -9 — but not yet deregistered.

killed by -9, not yet deregistered AND passing a health check. For some reason, we used to observe a couple of such instances in all our envs.

> Would you be also mind doing a curl <insert-base-url>/v1/health/checks/consul-esm and return the output for your zombie instance? For example, here’s my dead instance:

I can not, see my _UDPATE 2_ from the previous message.
We had couple of such "zombie" instances in every environment. They were just stuck there as healthy for an unknown reason for a very long time. After I noticed those log messages about ACL, I attempted deregistering those instances using a management-level token. This succeeded and we have never seen such instances again. As I said in that _UPDATE 2_, I tried the offerd reproducing scenario after that, and I was never able to get any "zombie" instances — as soon as I do `kill -9` consul-esm instance stops passing health check and recoordination happens as expected.

My only explanation is that those stuck instances were registered with the token, that got revoked (or received weaker privileges) later.

The issue I had with consul-esm is resolved.

Thank you very much for the follow-up and all the clarifications.

**Comment by lornasong** [comment]
@edio - got it, thanks for clarification! Looks like I misunderstood and thought the token-revoke as a separate issue. I understand now that you were considering token-revoke as the root cause for the zombie consul-esm instances you observed.

Great! Glad to hear that this consul-esm issue is resolved for you. Hopefully https://github.com/hashicorp/consul-esm/issues/53 captures the potential root cause of the zombie consul-esms you had. If not, feel free to reach out again.
